File: Semester3/Advanced Programming Methods/Labs/StudentsManagementAppC#/StudentsManagementApp/repository/AbstractRepository.cs
using System;
using System.Collections.Generic;
using StudentsManagementApp.domain.entities;
using StudentsManagementApp.domain.validator;

namespace StudentsManagementApp.repository
{
	public abstract class AbstractRepository<TId, TE>: ICrudRepository<TId, TE> where TE : class, IHasId<TId>
	{
		private readonly IValidator<TE> _validator;
		private readonly IDictionary<TId, TE> _map;

		/// <summary>
		/// Constructor for AbstractRepository class
		/// </summary>
		/// <param name="val">validator for TE entity</param>
		protected AbstractRepository(IValidator<TE> val)
		{
			_validator = val;
			_map = new Dictionary<TId, TE>();
		}
		
		public TE FindOne(TId id)
		{
			if (id == null)
			{
				throw new IllegalArgumentException("Id can't be null!");
			}
			return _map.ContainsKey(id) ? _map[id] : null;
		}

		public IEnumerable<TE> FindAll()
		{
			return _map.Values;
		}

		public virtual TE Save(TE entity)
		{
			if (entity == null)
			{
				throw new IllegalArgumentException("Entity can't be null");
			}
			_validator.Validate(entity);
			if (_map.ContainsKey(entity.Id))
			{
				return entity;
			}
			
			_map.Add(entity.Id, entity);
			//Console.WriteLine(entity);
			return null;
		}

		public virtual TE Delete(TId id)
		{
			if (id == null)
			{
				throw new IllegalArgumentException("Id can't be null!");
			}
			var found = FindOne(id);

			if (found != null)
			{
				_map.Remove(id);
			}

			return found;
		}

		public virtual TE Update(TE entity)
		{
			if (entity == null)
			{
				throw new IllegalArgumentException("Entity can't be null");
			}
			var found = FindOne(entity.Id);

			if (found == null)
			{
				return entity;
			}

			_map[entity.Id] = entity;
			return null;
		}
	}
}

File: Semester5/PPD/Lab5/src/repository/AbstractRepository.java
package repository;

import domain.entities.IEntity;

import java.util.HashMap;
import java.util.Map;

public class AbstractRepository<E extends IEntity<ID>, ID> implements IRepository<ID, E>{
    private Map<ID, E> map;
    public AbstractRepository(){
        map = new HashMap<>();
    }

    @Override
    public E findOne(ID id) {
        if(id == null){
            return null;
        }
        return map.get(id);
    }

    @Override
    public Iterable<E> findAll() {
        return map.values();
    }

    @Override
    public E save(E ent) {
        if(ent == null){
            return null;
        }
        E elem = findOne(ent.getId());
        map.put(ent.getId(), ent);
        return elem;
    }

    @Override
    public E delete(ID id) {
        if(id == null){
            return null;
        }
        E elem = findOne(id);
        if(elem == null){
            return null;
        }
        map.remove(id);
        return elem;
    }

    @Override
    public E update(E e) {
        E elem = map.get(e.getId());
        if(elem == null){
            return null;
        }
        this.save(e);
        return elem;
    }

    @Override
    public int size() {
        return map.values().size();
    }
}
